File: samples/clojure/api-sample-code/src/api_sample_code/create_issue_with_attachment.clj
(ns api-sample-code.create-issue-with-attachment
  (:require [org.httpkit.client :as http]
            [cheshire.core :as json]
            [pantomime.mime :refer [mime-type-of]]
            [clojure.string :as cs]))


(def HS-API-ENDPOINT "https://api.helpshift.com/v1/")


(defn construct-multipart-request-data
  "Constructs a multipart body given a data map containing attachment key.
  Input: {:email \"test@mail.com\"
          :message-body \"Create issue\"
          :attachment \"file-url\"}
  Output: [{:name \"email\" :content \"test@mail.com\"}
           {:name \"message-body\" :content \"Create issue\"}
           {:name \"attachment\" :filename \"file-name\" :content-type \"file-type\" :content \"java.io.File object\"}]"
  [data]
  (let [multipart-data (mapv (fn [k]
                               {:name (name k)
                                :content (k data)})
                             (keys (dissoc data :attachment )))
        attachment-file-path (:attachment data)
        multipart-data-with-attachment (conj multipart-data
                                             {:name "attachment"
                                              :filename (last (cs/split (:attachment data) #"/"))
                                              :content (clojure.java.io/file attachment-file-path)
                                              :content-type (mime-type-of attachment-file-path)})]
    multipart-data-with-attachment))


(defn create-issue
  "Create issue given a domain name, api key, issue data map."
  [domain api-key issue-data]
  (let [multipart-request-data (construct-multipart-request-data issue-data)
        create-issue-api-endpoint (str HS-API-ENDPOINT
                                       domain
                                       "/issues")
        api-response @(http/post create-issue-api-endpoint
                                 {:basic-auth api-key
                                  :multipart multipart-request-data})]
    (json/parse-string (:body api-response) true)))


(comment
  (create-issue "<DOMAIN>"
                "<API-KEY>"
                {:message-body "Test message"
                 :email "Test@mail.com"
                 :title "Issue title"
                 :tags (json/generate-string ["test"])
                 :meta (json/generate-string {:test 1})
                 :author-name "test-user"
                 :platform-type "web"
                 :app-id "<APP-ID>"
                 :attachment "resources/img_01.jpg"}))

File: samples/python/create_issue_with_attachment.py
import json
import requests
import mimetypes


DOMAIN = "<DOMAIN>"
API_KEY = "<API_KEY>"
ATTACHMENT_FILE_PATH = "<ATTACHMENT FILE PATH>"
TEST_PAYLOAD = {"email": "test@mail.com",
                "message-body": "Attachments Demo",
                "title": "Test issue title",
                "tags": json.dumps(["foo"]),
                "meta": json.dumps({"test": 1}),
                "author-name": "Test author",
                "platform-type": "web",
                "app-id": "random_app_20140423095931205-caae76e3be13ef6"}


def construct_attachment_object(attachment_url):
    file_name = attachment_url.split("/")[-1]
    file_type, encoding = mimetypes.guess_type(attachment_url)
    return {"attachment": (file_name,
                           open(attachment_url, "rb"),
                           file_type)}


def make_api_call(api_endpoint, api_key, payload, attachment={}):
    response = requests.post(api_endpoint,
                             auth=(api_key, ""),
                             data=payload,
                             files=attachment)
    print "API response status {0}".format(response.status_code)
    print "API response {0}".format(response.json())


if __name__ == "__main__":
    api_endpoint = 'https://api.helpshift.com/v1/{0}/issues'.format(DOMAIN)
    attachment_object = construct_attachment_object(ATTACHMENT_FILE_PATH)
    make_api_call(api_endpoint,
                  API_KEY,
                  TEST_PAYLOAD,
                  attachment_object)
